zebra-chain/src/mmr.rs

```diff
@@ -0,0 +1,43 @@
+//! Merkle mountain range structure that contains information about
+//! the block history as specified in ZIP-221.
+
+use std::borrow::Borrow;
+
+use crate::block::{Block, ChainHistoryMmrRootHash};
+
+/// Merkle mountain range structure.
+pub struct MerkleMountainRange {
+    // TODO
+}
+
+impl MerkleMountainRange {
+    pub fn push(&mut self, block: &Block) {
+        // TODO: zcash_history::Tree::append_leaf receives an Arc<Block>. Should
+        // change that to match, or create an Arc to pass to it?
+        todo!();
+    }
+
+    /// Extend the MMR with the given blocks.
+    // TODO: add Sapling and Orchard roots to parameters
+    pub fn extend<C>(&mut self, _vals: &C)
+    where
+        C: IntoIterator,
+        C::Item: Borrow<Block>,
+        C::IntoIter: ExactSizeIterator,
+    {
+        // TODO: How to convert C to `impl Iterator<Item = (Arc<Block>, sapling::tree::Root)>`
+        // for zcash_history::Tree::append_leaf_iter? Should change that to match C?
+        todo!();
+    }
```

Comment:
Let's `impl Extend for MerkleMountainRange` instead:
https://doc.rust-lang.org/std/iter/trait.Extend.html

This might be easier if we also `impl Extend for zebra_chain::primitives::zcash_history::Tree` as well. Feel free to make whatever changes are needed to the `Tree` API to make that work more easily.

Reply:
`Chain.blocks.values().map(|prepared_block| prepared_block.block)` should get you an iterator over all the `Arc<Block>`s you need. Since they came from a `BTreeMap<Height, _>`, they'll also be in the correct order.

Reply:
The issue with `Extend` is that the underlying `append_leaf` can return an error, so I don't think it will possible to use it. (The code right now is using it, because I'm `expect`ing that error, but will need to change)

Ah, thanks! I was wondering how to "convert" between iterators, had forgot about `map` :+1:

Reply:
You could use `try_fold` rather than `Extend`:
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_fold

But we can go with whatever works for now.